### PR TITLE
[#6856] Add ethereum/sign event

### LINKED ISSUE #6856: Add ethereum/sign event
### User Story

As a developer, I want an event to sign data. See https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_sign

```clojure
[ethereum/sign {:address "" :value "" :on-result [on-sign-result]}] ;; address should be optional
```

### Description

*Type*: Feature

### PR DESCRIPTION
**Pull request by tbenr**
fixes #6856

### CODE REVIEW

**Comment by jeluard** [comment]
@tbenr Looks great! Why is that a WIP ? :)

**Review by jeluard — APPROVED** [review]

**Comment by tbenr** [comment]
@jeluard because I haven't tested yet.. :)

**Comment by tbenr** [comment]
i get this error:

```
:code -32601
:message "The method eth_sign does not exist/is not available"
```

calling:

```
{"jsonrpc":"2.0","method":"eth_sign","params":["0x5541e3be81b76d76cdbf968516caa5a5b773763b","0xdeadbeaf"]}
```

I need to investigate more.. (wallet needs to be unlocked?)

File: src/status_im/constants.cljs [hunk 1]
```diff
@@ -200,6 +200,7 @@
 (def ^:const web3-send-transaction "eth_sendTransaction")
 (def ^:const web3-personal-sign "personal_sign")
 (def ^:const web3-get-logs "eth_getLogs")
+(def ^:const web3-sign "eth_sign")
```
Comment (flexsurfer) [inline src/status_im/constants.cljs]: eth_sign is deprecated, but anyway this issue should be solved in different way, we need to open wallet screen directly, added a comment

**Review by flexsurfer — CHANGES_REQUESTED** [review]

**Comment by flexsurfer** [comment]
signing process should be done same as send transaction, so we need to open signing screen (modal wallet) , https://github.com/status-im/status-react/blob/develop/src/status_im/extensions/ethereum.cljs#L61 , should be something like `open-modal-wallet-for-signing-message` , here you can find current implementation https://github.com/status-im/status-react/blob/72e7ae2fffe65639b6e2c13e8929dd42a643c41f/src/status_im/ui/screens/wallet/send/events.cljs#L154

File: src/status_im/constants.cljs (hunk 1 above)
Comment (tbenr) [reply src/status_im/constants.cljs]: now I get 

```
:code 32700
:message "method is unsupported by RPC interface"
```

Comment (flexsurfer) [reply src/status_im/constants.cljs]: sorry, i edited my comment

**Comment by jeluard** [comment]
@tbenr Sorry for the confusion! Does it make sense now?

**Comment by tbenr** [comment]
@jeluard yes, going to do that!

**Comment by tbenr** [comment]
@jeluard @flexsurfer it works to me...

**Comment by jeluard** [comment]
@tbenr Do you have an example of its usage and the kind of data it returns?

**Comment by tbenr** [comment]
@jeluard yes, but i realize i don't see any message in the signing dialog.. let me verify

**Comment by tbenr** [comment]
ok now it is ok.
that's the idea:

usage 1: plain text message
```
[ethereum/sign {:message "test message"
                  :on-result [store-result]}]
```

usage 2: raw data

```
[ethereum/sign {:data "0x617364"
                  :on-result [store-result]}]
```
you get result in `:result` or error message in `:error`.

it is possible to specify `:address` too (ens resolution is applied) but i don't know if makes sense in our context (can we have multiple account unlocked?)

File: src/status_im/extensions/ethereum.cljs [hunk 2]
```diff
@@ -204,3 +206,25 @@
            registry (get ens/ens-registries chain)]
        (ens/get-addr web3 registry name #(re-frame/dispatch (on-result {:result %}))))
      (re-frame/dispatch (on-result {:error (str "'" name "' is not a valid name")})))))
+
+;; EXTENSION SIGN -> SIGN MESSAGE
+(defn  execute-extension-sign [db {:keys [address message data id on-result]}]
```
Comment (jeluard) [inline src/status_im/extensions/ethereum.cljs]: What happens if the address is not the current wallet one?

File: src/status_im/extensions/core.cljs [hunk 3]
```diff
@@ -444,6 +444,13 @@
                  :value       :extensions/ethereum-transaction-receipt
                  :arguments   {:value     :string
                                :on-result :event}}
+                'ethereum/sign
+                {:permissions [:read]
+                 :value       :extensions/ethereum-sign
+                 :arguments   {:address?   :string
+                               :message?   :string
```
Comment (jeluard) [inline src/status_im/extensions/core.cljs]: I guess this implies we would need some way to express either this argument or this one :)

Comment (tbenr) [reply src/status_im/extensions/core.cljs]: true... when i was playing around with that i see there is a way to do something similar:

```
:arguments  {
                     :value  {:one-of #{
                                  {:message :string}
                                  {:data :string}}}
```

then we have to call

```
[ethereum/sign {:value {:data "0x617364"}}
                  :on-result [store-result]}]

[ethereum/sign {:value {:message "text"}}
                  :on-result [store-result]}]
```

are there other ways?

File: src/status_im/extensions/ethereum.cljs (hunk 2 above)
Comment (tbenr) [reply src/status_im/extensions/ethereum.cljs]: schermata 2018-11-30 alle 08 02 27

I think makes sense to remove address parameter, unless in some cases we have more then one account unlocked

**Comment by jeluard** [comment]
Right let’s remove it for now.

**Comment by jeluard** [comment]
Nahh it’s perfect as it is now !

**Comment by tbenr** [comment]
> Nahh it’s perfect as it is now !

I added a runtime check ;-) and removed :address

File: src/status_im/extensions/ethereum.cljs [hunk 4]
```diff
@@ -204,3 +206,22 @@
            registry (get ens/ens-registries chain)]
        (ens/get-addr web3 registry name #(re-frame/dispatch (on-result {:result %}))))
      (re-frame/dispatch (on-result {:error (str "'" name "' is not a valid name")})))))
+
+;; EXTENSION SIGN -> SIGN MESSAGE
+(handlers/register-handler-fx
```
Comment (flexsurfer) [inline src/status_im/extensions/ethereum.cljs:211]: reformatted little bit according to our code guides 


```
(handlers/register-handler-fx
 :extensions/ethereum-sign
 (fn [{db :db :as cofx} [_ _ {:keys [message data id on-result]}]]
   (if (= (nil? address) (nil? data)) 
      {:dispatch (on-result {:error "only one of :message and :data can be used"}))
     (fx/merge cofx
      {:db (assoc-in db [:wallet :send-transaction]
                       {:id                id
                        :from             (ethereum/normalized-address  (get-in db [:account/account :address])))
                        :data             (or data (str "0x" (abi-spec/from-utf8 message)))
                        :on-result        [:extensions/wallet-ui-on-result on-result]
                        :on-error         [:extensions/wallet-ui-on-error on-result]
                        :method           constants/web3-personal-sign})}
      (navigation/navigate-to-cofx :wallet-sign-message-modal nil)))))
```

**Review by flexsurfer — APPROVED** [review]
thanks @tbenr  great job!

**Comment by asemiankevich** [comment]
@jakubgs what is the build generated here for e2e? https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-181130-092726-0169dc-pr.sha256 cc @antdanchenko @hetvart 

<blockquote></blockquote>

**Comment by jakubgs** [comment]
>@jakubgs what is the build generated here for e2e?

Not sure what you are asking. Did you mean to ask "why" instead of "what"? Because if you wanted to ask "what", it's clearly and android end-to-end build.

**Comment by asemiankevich** [comment]
@jakubgs i wanted to ask why the e2e build has `pr.sha256` ending instead of `apk`. Sorry for my broken english

**Comment by jakubgs** [comment]
Oooooh, I didn't notice that, the link is indeed incorrect. Let me check that.

**Comment by jakubgs** [comment]
Thanks for reporting this @asemiankevich, there was indeed a small bug in the change that introduced SHA256 checksums. Fix is here: https://github.com/status-im/status-react/pull/6968
Sorry for misunderstanding, I didn't know you were referring to the github comment with the links.

**Comment by asemiankevich** [comment]
@jakubgs no worries Jacob, thanks for fixing this so fast!

**Pull request merged**